File: Week 3/id_445/55.跳跃游戏.cpp
/*
 * @lc app=leetcode.cn id=55 lang=cpp
 *
 * [55] 跳跃游戏
 */
#include <vector>
using namespace std;
// @lc code=start
class Solution {
public:
//solution 1 
    // bool canJump(vector<int>& nums) {
    //     int maxStep=0;
        
    //     for (int i = 0; i < nums.size(); i++)
    //     {
    //         if(i>maxStep) return false;
    //         maxStep=max(maxStep,i+nums[i]);
    //     }
    //     return true;
        
    // }
//solution 2 beats 99.92%
        bool canJump(vector<int>& nums) {
        int maxStep=0;
        
        for (int i = 0; i < nums.size(); i++)
        {
            if(i>maxStep) return false;
            if(i+nums[i]>nums.size() ) return true;
            maxStep=max(maxStep,i+nums[i]);
        }
        return true;
        
    }
};
// @lc code=end

File: Week 3/id_520/55.跳跃游戏.java
/*
 * @lc app=leetcode.cn id=55 lang=java
 *
 * [55] 跳跃游戏
 */

// @lc code=start
class Solution {
    public boolean canJump(int[] nums) {
        
        //输入参数为无解
        if (nums == null) {
            return false;
        }

        //能够到达的最后位置的下标，初始值是最后一个元素的位置，意为，最后一个元素能够到达最后一个地方（从后往前）
        int canReachable = nums.length - 1;
        for (int i=canReachable; i>=0; i--) {
            //若当前下标加上下标所在的步数可以到达canReachable，那么就后退到那个位置
            if (nums[i]+i >= canReachable) {
                canReachable = i;
            }
        }
        return canReachable == 0;
    }
}
// @lc code=end

